nixos/mattermost: add preferNixConfig option, plugin support, and tests

numinit opened the pull request:
###### Motivation for this change

One use case for Mattermost configuration is doing a "mostly
mutable" configuration where NixOS module options take priority
over Mattermost's config JSON.

Add a `preferNixConfig` option that prefers configured Nix options
over what's configured in Mattermost config if mutableConfig is set.

Add a `plugins` option that should be a list of Nix store paths to
Mattermost .tar.gz plugin bundles. Take it into account when building
the Mattermost config.

Remove the reliance on readFile (it's flake incompatible) and use
jq instead.

Merge Mattermost configs together on Mattermost startup, depending
on configured module options.

Write tests for mutable, mostly mutable, and immutable configurations.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

numinit committed:
nixos/mattermost: add preferNixConfig option and tests

One use case for Mattermost configuration is doing a "mostly
mutable" configuration where NixOS module options take priority
over Mattermost's config JSON.

Add a preferNixConfig option that prefers configured Nix options
over what's configured in Mattermost config if mutableConfig is set.

Remove the reliance on readFile (it's flake incompatible) and use
jq instead.

Merge Mattermost configs together on Mattermost startup, depending
on configured module options.

Write tests for mutable, mostly mutable, and immutable configurations.

numinit committed:
nixos/mattermost: Support declarative Mattermost plugins

numinit committed:
nixos/mattermost: Simplify plugin derivation build, improve startup perf

numinit commented:
Should resolve #111076.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/531

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/543

happysalada commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -6,23 +6,97 @@ let
+  '';
+
+  mattermostPluginDerivations = with pkgs;
+    if cfg.plugins == null then null
```
how about making cfg.plugins default to an empty list so that you can get rid of the if condition all together?

numinit replied:
Good call. I had some gnarly code here before, this is a holdover.

mohe2015 replied:
@numinit Do you think you could find some time to fix this? I would love to see this merged as there seem some further efforts to improve this module.

numinit replied:
Sorry about the delay, life got in the way... pushing these changes, and another change that silences "read only file system" warnings at startup.

happysalada commented:
Generally looks good. I have one tiny nit.
On a separate note a couple of things (optional, just if you are interested)
- mattermost hasn't been updated in a while.
- There is a PR to split it into a server and the webapp. (just in case you have particular knowledge about mattermost and are interested). https://github.com/NixOS/nixpkgs/pull/126629

mohe2015 reviewed (commented):
Behaviour-wise LGTM! Code-wise too but didn't think about the edge cases (mutable, preferNixConfig) that I didn't use too much. Also fixed the flake-issue I had (expectedly).

Tests failed though (rebased to nixos-unstable-small):
```
vm-test-run-mattermost> Fatal Python error: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
vm-test-run-mattermost> Python runtime state: finalizing (tstate=0x434840)
vm-test-run-mattermost> Thread 0x00007ffff5177640 (most recent call first):
vm-test-run-mattermost>   File "/nix/store/hpsyw7zpjwq9g756kq6xk5k0xbdr8xzn-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 807 in process_serial_output
vm-test-run-mattermost> Thread 0x00007ffff5978640 (most recent call first):
vm-test-run-mattermost>   File "/nix/store/hpsyw7zpjwq9g756kq6xk5k0xbdr8xzn-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 807 in process_serial_output
vm-test-run-mattermost> Thread 0x00007ffff6179640 (most recent call first):
vm-test-run-mattermost>   File "/nix/store/hpsyw7zpjwq9g756kq6xk5k0xbdr8xzn-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 99 in eprint
vm-test-run-mattermost>   File "/nix/store/hpsyw7zpjwq9g756kq6xk5k0xbdr8xzn-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 184 in log_serial
vm-test-run-mattermost>   File "/nix/store/hpsyw7zpjwq9g756kq6xk5k0xbdr8xzn-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 342 in log_serial
vm-test-run-mattermost>   File "/nix/store/hpsyw7zpjwq9g756kq6xk5k0xbdr8xzn-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 811 in process_serial_output
vm-test-run-mattermost> Current thread 0x00007ffff784af80 (most recent call first):
vm-test-run-mattermost> <no Python frame>
vm-test-run-mattermost> /nix/store/qg8qhrxiab3r87xmaxbq565g1g8bnl57-stdenv-linux/setup: line 1280:     7 Aborted                 (core dumped) LOGFILE=/dev/null tests='exec(os.environ["testScript"])' /nix/store/ichqla4xnx2fh9x7jfwxp9g64kldmaa1-nixos-test-driver-mattermost/bin/nixos-test-driver
```
EDIT: they worked later, maybe flaky, maybe I rebased master idk

numinit committed:
nixos/mattermost: silence startup warnings from chmod/chown; improve API

mohe2015 commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -44,7 +43,7 @@ let
 
   mattermostPlugins = with pkgs;
-    if mattermostPluginDerivations == null then null
+    if mattermostPluginDerivations == [] then null
```
Didn't test it but could this condition not be removed? (At the cost of building a probably empty derivation). Or does this not allow manual plugin install or something else then?

mohe2015 replied:
As far as I understand the code this is probably the reason.

numinit replied:
Manual plugin installs don't work for plugins that have binaries, since the dynamic linker may not be set properly in them, and we have to run autoPatchelf as part of the derivation build to make them work. That being said, I think there's a way to support both manual and Nix-installed plugins if we symlink in all the plugins when the server starts

mohe2015 replied:
> That being said, I think there's a way to support both manual and Nix-installed plugins if we symlink in all the plugins when the server starts
I don't think we need to implement this for now. Could always be done later.

mohe2015 commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -77,6 +157,16 @@ in
         '';
       };
 
+      preferNixConfig = mkOption {
```
I'm not sure if I like this. It seems like this just overrides the config at server startup so you can still configure things differently in the web UI but you will suddenly loose settings then. I understand that you actively need to enable it but I don't know if we really need/want this except maybe If you are using this.

numinit replied:
You used to lose all the settings if you disabled mutable configs entirely. This provides an intermediate option.

mohe2015 commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -6,23 +6,96 @@ let
+        PluginSettings = {
+          Enable = true;
+          EnableUploads = false;
+          Directory = "${cfg.statePath}/plugins/server";
```
I think at least these two options should always be set as otherwise you can't install plugins even if mutableConfig is true as the plugin directory is read only.

mohe2015 replied:
Alternatively the paths could be changed to the official ones.

numinit replied:
Good call - I'll look at how to make the plugins directory look more "normal."

numinit replied:
I don't think we can. The client directory is symlinked into the nix store. So it has to be plugins/server and plugins/client instead. Not a huge deal though, we can just use those paths.

numinit committed:
mattermost: 5.37.5 -> 6.2.1

Add overridable arguments:
- versionSuffix (default: nixpkgs)
- buildDate (default: 1970-01-01)
- storePathAsBuildHash (default: false)

numinit committed:
mattermost: update release notes

numinit committed:
nixos/mattermost: update service for 6.0+

numinit commented:
See #152899 for the Mattermost updates.

numinit commented:
Apologies for the delay, this branch is now 6.x-compatible, and rebased against #152899.

happysalada commented:
Thanks a lot for this. Just running a quick smoke test before merging.

happysalada commented:
Result of `nixpkgs-review pr 119200` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mattermost</li>
  </ul>
</details>

ryantm commented:
@numinit I would appreciate it if you could tag me when making the decision of switching from the Extended Support Release versions to the latest release. My experience with Mattermost has been there they do not always handle database migrations between versions well, so I had chosen to only support the ESR versions in hopes that they would be more likely not to break.

That said, I appreciate your work on Mattermost, and maybe it could be fine to give them another chance at upgrade compatibility.

I also noticed you added back in that version suffix of -nixpkgs to the server derivation. I had removed that because I think there was some warning in the server logs on startup relating to it. I think I remember that it was trying to see if the version of Mattermost is out of date and failing to detect what version it was.

happysalada commented:
I merged a little fast here as I didn't want this to go to waste. I wasn't sure if you were going to be available @ryantm , I'll make sure to give reasonable time for feedback next time there is a contribution on mattermost.

ryantm commented:
@happysalada I think you were fine to merge it. This PR has been open for 9 months, and I wasn't engaging with it. I appreciate both your help!

numinit commented:
@ryantm

Good to know. Thanks, will do next time.

Nothing in the module is really 6.x _specific_ - I was able to change the version to 5.37.6 (latest ESR) and the smoke tests pass just fine. It's fine with me if we revert Mattermost to v5.37.6 and add a mattermost_6 package but still keep v5 as the default then. [Here's a PR](https://github.com/NixOS/nixpkgs/pull/153141) that does it.

re: the version - It doesn't complain if you suffix the version rather than prefixing it. I saw the same warning, and it was fixed by changing it to a suffix.

numinit commented:
As a followup, I didn't have any trouble going 5.37 to 6 on a fairly large instance, so it should be all right. But, still, best to be safe and prefer the LTS release.
